### PR TITLE
Add logical form name to schema metadata [V2]

### PR DESCRIPTION
**Pull request by amitjoshi438**
This pull request mainly involves the addition of a new property `logicalFormName` to the `FileData` class and related methods in the `WebExtensionContext` and `FileDataMap` classes. The changes also include modifications in the `remoteFetchProvider.ts` file to accommodate this new property and update the metadata accordingly.

Addition of `logicalFormName`:

* [`src/web/client/context/fileData.ts`](diffhunk://#diff-6ecb7ca7e2966d7538f8748f3308c75f7694bb884c81446eb0d6fae23bf2edf6R33): Added a new private property `_logicalFormName` to the `FileData` class. Corresponding getter and setter methods were also added. The property was also included in the class constructor. [[1]](diffhunk://#diff-6ecb7ca7e2966d7538f8748f3308c75f7694bb884c81446eb0d6fae23bf2edf6R33) [[2]](diffhunk://#diff-6ecb7ca7e2966d7538f8748f3308c75f7694bb884c81446eb0d6fae23bf2edf6R74-R77) [[3]](diffhunk://#diff-6ecb7ca7e2966d7538f8748f3308c75f7694bb884c81446eb0d6fae23bf2edf6R93-R96) [[4]](diffhunk://#diff-6ecb7ca7e2966d7538f8748f3308c75f7694bb884c81446eb0d6fae23bf2edf6L98-R108) [[5]](diffhunk://#diff-6ecb7ca7e2966d7538f8748f3308c75f7694bb884c81446eb0d6fae23bf2edf6R122)

* [`src/web/client/WebExtensionContext.ts`](diffhunk://#diff-e1e4b1cfb30bca61f70c080e4150842da2a4f1837346604bc23ac58a820ad1caL405-R406): The `setEntity` method of the `WebExtensionContext` class was updated to include `logicalFormName` as a parameter. [[1]](diffhunk://#diff-e1e4b1cfb30bca61f70c080e4150842da2a4f1837346604bc23ac58a820ad1caL405-R406) [[2]](diffhunk://#diff-e1e4b1cfb30bca61f70c080e4150842da2a4f1837346604bc23ac58a820ad1caL418-R420)

* [`src/web/client/context/fileDataMap.ts`](diffhunk://#diff-fb3567e1b488a6ac763a65e05fe58ceb93804293e1f6c9b7028e919098b47464L28-R29): The `setEntity` method of the `FileDataMap` class was updated to include `logicalFormName` as a parameter. [[1]](diffhunk://#diff-fb3567e1b488a6ac763a65e05fe58ceb93804293e1f6c9b7028e919098b47464L28-R29) [[2]](diffhunk://#diff-fb3567e1b488a6ac763a65e05fe58ceb93804293e1f6c9b7028e919098b47464L40-R42)

Updates in `remoteFetchProvider.ts`:

* [`src/web/client/dal/remoteFetchProvider.ts`](diffhunk://#diff-3bdc6b0be6c5885d51a3c8dea2d155cd6c1ab149514f5f75b477a401245493feL17-R17): The `createFile` and `createVirtualFile` functions were updated to include `logicalFormName` as a parameter. The `createFile` function was also updated to fetch metadata values for `logicalEntityName` and `logicalFormName` and pass them to `createVirtualFile`. [[1]](diffhunk://#diff-3bdc6b0be6c5885d51a3c8dea2d155cd6c1ab149514f5f75b477a401245493feL17-R17) [[2]](diffhunk://#diff-3bdc6b0be6c5885d51a3c8dea2d155cd6c1ab149514f5f75b477a401245493feR27) [[3]](diffhunk://#diff-3bdc6b0be6c5885d51a3c8dea2d155cd6c1ab149514f5f75b477a401245493feR435) [[4]](diffhunk://#diff-3bdc6b0be6c5885d51a3c8dea2d155cd6c1ab149514f5f75b477a401245493feR458-R460) [[5]](diffhunk://#diff-3bdc6b0be6c5885d51a3c8dea2d155cd6c1ab149514f5f75b477a401245493feL471-R477) [[6]](diffhunk://#diff-3bdc6b0be6c5885d51a3c8dea2d155cd6c1ab149514f5f75b477a401245493feR629) [[7]](diffhunk://#diff-3bdc6b0be6c5885d51a3c8dea2d155cd6c1ab149514f5f75b477a401245493feL637-R645)

Other changes:

* [`src/web/client/schema/constants.ts`](diffhunk://#diff-c49e75521c24e65f4c0cde53b75034aba44907b480d222eb09c807046e57de56R38): Added a new key `DATAVERSE_FORM_NAME` to `schemaMetaDataKey` enum.

* [`src/web/client/schema/portalSchema.ts`](diffhunk://#diff-d2f0f5db4d90218c70bc6005e26d85bae0d8997de232d1f2fd62d35a4a22863bL382-R382): Updated `_dataverseentitymetadata` to include `content.formname` for `_dataverseformname`.

* [`src/web/client/utilities/fileAndEntityUtil.ts`](diffhunk://#diff-02b9a80bec87b6e3c39911c83e2d679402dbe0c4e87ceddf6bf13bd5f3b6846cR46-R49): Added a new function `getFormLogicalName` to fetch the logical form name for a given file path.

* [`src/web/client/utilities/schemaHelperUtil.ts`](diffhunk://#diff-a62e9afcf7842b91b4f185c56482491466aed38cebe061b82391d9c9a57fff79R30-R34): Added a new function `getLogicalFormNameParameter` to fetch the logical form name parameter for a given entity.

* [`src/web/client/utilities/urlBuilderUtil.ts`](diffhunk://#diff-0edb413159c2e0d7b6c0f63ee1b67cc0864dd3f3df93139f80629468a969a7f9R219-R237): Added a new function `getMetadataInfo` to fetch metadata values for given metadata keys.

### CODE REVIEW

File: src/web/client/context/fileData.ts [hunk 1]
```diff
@@ -95,7 +104,8 @@ export class FileData implements IFileData {
         encodeAsBase64?: boolean,
         mimeType?: string,
         isContentLoaded?: boolean,
-        logicalEntityName?: string
+        logicalEntityName?: string,
+        logicalFormName?: string
```
Comment (tyaginidhi) [inline src/web/client/context/fileData.ts:108]: we should not make this another parameter here - as its not generic parameter used by all entities - instead this should go in an interface with entityMetadata that can be expanded later - keep this one as a nullable param in that interface.

File: src/web/client/dal/remoteFetchProvider.ts [hunk 2]
```diff
@@ -453,6 +452,10 @@ async function createFile(
         fileContent = getAttributeContent(result, attributePath, entityName, entityId);
     }
 
+    const metadataKeys = getEntityParameters(entityName);
+    const metadataValues = getMetadataInfo(result, metadataKeys.filter(key => key !== undefined) as string[]);
+    console.log("Metadata entity name: "+ metadataValues.logicalEntityName)
```
Comment (tyaginidhi) [inline src/web/client/dal/remoteFetchProvider.ts:457]: stray console.log

File: src/web/client/dal/remoteFetchProvider.ts [hunk 3]
```diff
@@ -620,8 +624,10 @@ async function createVirtualFile(
     isPreloadedContent?: boolean,
     mappingEntityId?: string,
     logicalEntityName?: string,
+    entityMetadata?: SchemaEntityMetadata,
     rootWebPageId?: string,
 ) {
+    console.log("Metadata: "+ entityMetadata?.logicalFormName)
```
Comment (tyaginidhi) [inline src/web/client/dal/remoteFetchProvider.ts:630]: stary console.log

File: src/web/client/context/fileDataMap.ts [hunk 4]
```diff
@@ -25,7 +26,8 @@ export class FileDataMap {
         isBase64Encoded: boolean,
         mimeType?: string,
         isContentLoaded?: boolean,
-        logicalEntityName?: string
+        logicalEntityName?: string,
+        entityMetadata?: SchemaEntityMetadata
```
Comment (tyaginidhi) [inline src/web/client/context/fileDataMap.ts:30]: can be named better - can stay consistent with entityMetadata in name

File: src/web/client/dal/remoteFetchProvider.ts [hunk 5]
```diff
@@ -634,7 +640,8 @@ async function createVirtualFile(
         encodeAsBase64,
         mimeType,
         isPreloadedContent,
-        logicalEntityName
+        logicalEntityName,
```
[the remark this answers is not in the document]
Comment (amitjoshi438) [reply src/web/client/dal/remoteFetchProvider.ts:641]: I felt it's common to every entity and currently passed separately.

[the remark this answers is not in the document]
Comment (amitjoshi438) [reply src/web/client/dal/remoteFetchProvider.ts]: Reviewed it again and it made sense to pass it as part of metadata

File: src/web/client/utilities/urlBuilderUtil.ts [hunk 6]
```diff
@@ -216,6 +217,34 @@ export function getLogicalEntityName(result: any, logicalEntityName?: string) {
...
+
+    if (metadataKeys) {
+        for (const key of metadataKeys) {
+            const attributePath = getAttributePath(key);
+            const value = attributePath.relativePath.length > 0
+                ? JSON.parse(result[attributePath.source])[attributePath.relativePath]
+                : result[attributePath.source];
+
+            switch (key) {
+                case 'content.entityname':
```
Comment (tyaginidhi) [inline src/web/client/utilities/urlBuilderUtil.ts:232]: nit: See if we can keep them as const list that can be expanded in future.

**Review by tyaginidhi — APPROVED** [review]

File: src/web/client/utilities/urlBuilderUtil.ts (hunk 6 above)
Comment (amitjoshi438) [reply src/web/client/utilities/urlBuilderUtil.ts]: Done

**Review by BidishaMS — APPROVED** [review]

**Pull request merged**